### ISSUE TITLE
b.external() role is unclear

### ISSUE THREAD

**Issue by jmm** [substack/node-browserify#1076]
The `b.external()` documentation refers to excluding *files*, but is it for:

A) Registering a *string* (module ID) that should be skipped during bundling

or

B) Registering a *file* that should be excluded

or

C) Is it supposed to be able to differentiate between the two

?

(If it's `B` or `C`, I've already done a lot of work on that, so please give me the opportunity to submit a pull request.)

Allow me to illustrate the difference:

Given that there is a file `/somedir/jquery.js` and this `/somedir/entry.js`:

```js
require('./jquery');
```

Should `jquery.js` be in the output of the following runs of `/somedir/bundler.js`?:

Scenario # 1

```js
browserify('./entry.js')
  .external('./jquery')
```

If `A` then `false`.
If `B` then `false`.

Scenario # 2

```js
browserify('./entry.js')
  .external('./jquery.js')
```

If `A` then `true`.
If `B` then `false`.

****************************************

Are there good use cases for doing?:

```js
// Leading './' in expose
browserify().require('whatever', {expose: './whatever'});

// Leading '../' in expose
browserify().require('whatever', {expose: '../whatever'});
```

**Comment by peter-mouland** [comment]
for me this is the same as `-e` or `-external` on the CLI.

perfect for creating vendor bundles that get used throughout your app i.e. vendor.js includes jQuery and some plugins.
app.js although includes require('jquery'), jquery wont actually get bundled.

what i am confused about is why there isn't a `opts.external` for the API??

**Comment by jmm** [comment]
> for me this is the same as -e or -external on the CLI.

API or CLI it's the same question: are the values passed for this just `===` compared to the strings passed in `require()` calls or does browserify attempt to resolve one or both values to a pathname? Actually I just noticed that the CLI documentation for it says `Files can be globs.`

> what i am confused about is why there isn't a opts.external for the API??

I'm afraid I don't understand your question. `opts.external` as opposed to `b.external()`?, which you noted is equivalent to `--external` for the CLI.

**Comment by peter-mouland** [comment]
yeah. I am not using the cli, so this doesn't help me, so i have to code around it :
```javascript
var b = browserify(options);
        if (options.external){
            b.external(options.external);
        }
        b.bundle(...)
```
rather than  `browserify(options).bundle(..)`;

**Comment by jmm** [comment]
I see. Where is `options.external` coming from in this case?

**Comment by peter-mouland** [comment]
just a browserify config object that a build-tool passes to the API e.g.
```javascript
    browserify: {
        insertGloabals : false,
        detectGlobals : false,
        noParse: [
            './bower_components/jquery/dist/jquery.js',
            './bower_components/d3/d3.js',
        ],
        external: [
            './bower_components/jquery/dist/jquery.js',
            './bower_components/d3/d3.js',
        ]
    }
```

**Comment by jmm** [comment]
Ok, well if you want that I suggest you open a new issue for it (after making a reasonable attempt to check that there isn't already an issue for it of course), because this issue is about something completely different.
